### DISCUSSION TITLE
Cairo profiler

### DISCUSSION THREAD

**Discussion by MaksymilianDemitraszek** [software-mansion/protostar#542] [Ideas]
Hi everybody. I am doing some research on gas optimisation tool for Cairo contracts (Cairo profiler) we will build in the near future as a part of Protostar.

It would provide detailed analysis of each function cost and how many times each function is used in the executed test, similar to Foundry(https://book.getfoundry.sh/forge/gas-reports). It would also include profiling with fuzz tests to help find edge cases.

I’m wondering what features would you expect from such a tool? Do you maybe know any interesting tooling we could get inspired by?

**Comment by amanusk** [comment]
An awesome feature to have is something like the tenderly flame graph for gas usage.
Flame graphs are quite a known took, and exist in go rust etc.

Ideally this is something that can be created and viewed locally in the browser. 

Here is an example from tenderly:

https://dashboard.tenderly.co/public/defisaver-v2/prod-simulations/fork-simulation/00892a16-a402-46a9-8f83-92a5750f455a/gas-usage

**Comment by MaksymilianDemitraszek** [comment]
We plan to implement first version of profiler as a `--profile` flag for test command.

We will use pprof as our interface because it supports lots of formats by default (graph, flame graph).

![image](https://user-images.githubusercontent.com/32527780/182375160-b9ba8789-a8b9-4bad-be5d-84fd02b93358.png)
![image](https://user-images.githubusercontent.com/32527780/182375177-829c2c2d-64c2-4a58-b2df-fd6289a25fb5.png)

**Comment by MaksymilianDemitraszek** [comment]
Updating about the progress, we are currently successfully managed to profile memholes. Currently I am working on the support for including called contracts into profile.

![Screenshot 2022-08-29 at 17 02 58](https://user-images.githubusercontent.com/32527780/187373388-965ecdea-f2b0-4ea0-92e6-093b36f64bcd.png)

**Comment by MaksymilianDemitraszek** [comment]
https://github.com/software-mansion/protostar/milestone/22
